### PR TITLE
Sea Turtles - Tamara B.

### PR DESCRIPTION
**Pull request by Myranae**

### CODE REVIEW

File: tests/unit_tests/test_wave_01.py [hunk 1]
```diff
@@ -49,7 +49,6 @@ def test_removing_not_found_is_false():
 
     result = vendor.remove(item)
 
-    raise Exception("Complete this test according to comments below.")
-    # *********************************************************************
-    # ****** Complete Assert Portion of this test **********
-    # *********************************************************************
+    assert result == False
+    assert len(vendor.inventory) == 3
+    assert vendor.inventory == ["a", "b", "c"]
```
Comment (anselrognlie) [inline tests/unit_tests/test_wave_01.py:54]: 👍

File: tests/unit_tests/test_wave_02.py [hunk 2]
```diff
@@ -34,7 +34,4 @@ def test_get_no_matching_items_by_category():
 
     items = vendor.get_by_category("electronics")
 
-    raise Exception("Complete this test according to comments below.")
-    # *********************************************************************
-    # ****** Complete Assert Portion of this test **********
-    # *********************************************************************
+    assert items == []
```
Comment (anselrognlie) [inline tests/unit_tests/test_wave_02.py:37]: 👍 

This always looks strange to me since I always want to write empty list checks as `not items`, but here in the tests we want to be specific like this. In application Python code, we should use the more permissive checks (`not items`), but in a test, we want to ensure that the thing returned really is a list and that it is empty.

File: tests/unit_tests/test_wave_06.py [hunk 3]
```diff
@@ -76,7 +76,15 @@ def test_swap_best_by_category():
         their_priority="Decor"
     )
 
-    raise Exception("Complete this test according to comments below.")
+    assert result
```
Comment (anselrognlie) [inline tests/unit_tests/test_wave_06.py:79]: Much like the empty list check in wave 2, here, we may want to perform an explicit check with `True` to guarantee the specified behavior of returning `True`, not just truthiness.

File: tests/unit_tests/test_wave_06.py [hunk 4]
```diff
@@ -76,7 +76,15 @@ def test_swap_best_by_category():
...
+    assert len(tai.inventory) == 3
+    assert len(jesse.inventory) == 3
+    assert tai.inventory == [item_a, item_b, item_f]
+    assert jesse.inventory == [item_d, item_e, item_c]
```
Comment (anselrognlie) [inline tests/unit_tests/test_wave_06.py:83]: The specification doesn't stipulate where the swapped items will end up in the output. Consider using `in` checks to avoid implying an ordering.

File: tests/unit_tests/test_wave_06.py [hunk 5]
```diff
@@ -109,7 +117,15 @@ def test_swap_best_by_category_reordered():
...
     )
 
-    raise Exception("Complete this test according to comments below.")
+    assert result
+    assert len(tai.inventory) == 3
+    assert len(jesse.inventory) == 3
+    assert tai.inventory == [item_b, item_a, item_f]
+    assert jesse.inventory == [item_e, item_d, item_c]
+    assert item_c not in tai.inventory
+    assert item_f not in jesse.inventory
```
Comment (anselrognlie) [inline tests/unit_tests/test_wave_06.py:126]: Similar considerations here as for `test_swap_best_by_category`

File: tests/unit_tests/test_wave_06.py [hunk 6]
```diff
@@ -194,7 +210,17 @@ def test_swap_best_by_category_no_match_is_false():
         their_priority="Clothing"
     )
 
-    raise Exception("Complete this test according to comments below.")
+    assert not result
```
Comment (anselrognlie) [inline tests/unit_tests/test_wave_06.py:213]: 👍

File: tests/unit_tests/test_wave_06.py [hunk 7]
```diff
@@ -227,7 +253,17 @@ def test_swap_best_by_category_no_other_match_is_false():
         their_priority="Decor"
     )
 
-    raise Exception("Complete this test according to comments below.")
+    assert not result
```
Comment (anselrognlie) [inline tests/unit_tests/test_wave_06.py:256]: 👍

File: tests/unit_tests/test_enhancements.py [hunk 8]
```diff
@@ -0,0 +1,141 @@
...
+    tai = Vendor(
+        inventory=[item_a, item_b, item_c, item_d]
+    )
+
+    # Act
+    result = tai.get_newest_item()
+
+    assert result
+    assert len(tai.inventory) == 4
+    assert tai.inventory == [item_a, item_b, item_c, item_d]
```
Comment (anselrognlie) [inline tests/unit_tests/test_enhancements.py:24]: When there's no real expectation of something being edited, we don't really need to check that it hasn't been. This function is closer in operation to `get_best_by_category` than `swap_best_by_category`. Notice the tests for the former don't check that the inventory hasn't been modified.

File: tests/unit_tests/test_enhancements.py [hunk 9]
```diff
@@ -0,0 +1,141 @@
...
+        inventory=[item_e, item_f, item_g]
+    )
+
+    result = tai.swap_by_newest(jesse)
+
+    assert not result
+    assert len(jesse.inventory) == 3
+    assert len(tai.inventory) == 4
+    assert jesse.inventory == [item_e, item_f, item_g]
+    assert tai.inventory == [item_a, item_b, item_c, item_d]
```
Comment (anselrognlie) [inline tests/unit_tests/test_enhancements.py:97]: For `swap_by_newest`, this _is_ where we would want to check that things have or have not changed, since this function has a reasonable expectation of modifying the inventories. Again, we might prefer `in` checks over requiring a certain swap behavior.

File: swap_meet/vendor.py [hunk 10]
```diff
@@ -1,2 +1,158 @@
...
+        ---Returns---
+        Returns False if there were no items of category my_priority in other's inventory or if there were no items of category their_priority in the current instance's inventory. Returns true if these exist and were swapped.
+        '''
+        to_trade = self.get_best_by_category(their_priority)
+        to_receive = other.get_best_by_category(my_priority)
+        if not to_trade or not to_receive:
+            return False
+        return self.swap_items(other, to_trade, to_receive)
+
+    def get_newest_item(self) -> Item:
```
Comment (anselrognlie) [inline swap_meet/vendor.py:135]: I almost put a comment at the top that you didn't need to import `Item` in `Vendor`, but because of this type hint, you do! 😄 

This turns out to be sort of a weird project to try out type-hinting on. The later wave stuff all works on Item instances, but wave 1 was using strings as the inventory contents, so you end up having a bit of an identity crisis for what to use for the annotation.

If you do decide to keep working with type-hints, be sure to try running your code through a [type checker](https://www.infoworld.com/article/3575079/4-python-type-checkers-to-keep-your-code-clean.html) to see what they tell you about your annotations. Python itself ignores them, so they primarily gain value when we use a tool to help us find places where we're violating our type info.

File: swap_meet/vendor.py [hunk 11]
```diff
@@ -1,2 +1,158 @@
...
-    pass
\ No newline at end of file
+    '''
+    A class to represent a vendor at the swap meet.
+
+    ...
+
+    ---Attributes---
+    inventory : list, optional
+        None by default, else a list of the vendor's items
```
Comment (anselrognlie) [inline swap_meet/vendor.py:12]: The `inventory` attribute itself is never `None` and not optional. It's always a list. The constructor parameter, on the other hand, _is_ optional.

File: swap_meet/vendor.py [hunk 12]
```diff
@@ -1,2 +1,158 @@
...
+        inventory : list, optional
+            a list of item objects (default is None)
+        '''
+        if not inventory:
+            self.inventory = []
+        else:
+            self.inventory = inventory
+    
+    def add(self, one_item):
+        '''Adds one_item to the instance's inventory list'''
```
Comment (anselrognlie) [inline swap_meet/vendor.py:48]: This is honestly about the level of docstring that I can live with! 😆 

You're doing great with docstrings, but opinions about how much to use them vary wildly from team to team. So feel free to use docstrings however they provide the most use to you, but be ready to adapt your documentation to match the prevailing style of any team you join (and expect what it should look like to be a constant battle!). Seriously, people get really worked up about this. 😆 

Even in languages that don't officially support the notion of a docstring, a team will often have a tool to extract documentation from structured comments, so this is something that extends well beyond Python!

File: swap_meet/vendor.py [hunk 13]
```diff
@@ -1,2 +1,158 @@
...
+        Constructs all the necessary attributes for the vendor object
+
+        ---Paramters---
+        inventory : list, optional
+            a list of item objects (default is None)
+        '''
+        if not inventory:
+            self.inventory = []
+        else:
+            self.inventory = inventory
```
Comment (anselrognlie) [inline swap_meet/vendor.py:45]: Nice job handling the default inventory case. One thing to notice is that you might want to explicitly check for `None` rather than falsy, since written this way, if the caller explicitly passed in `[]` (their own empty list), this logic would still replace the passed in empty list with the new local empty list.

File: swap_meet/vendor.py [hunk 14]
```diff
@@ -1,2 +1,158 @@
...
+    
+    def add(self, one_item):
```
Comment (anselrognlie) [inline swap_meet/vendor.py:47]: 👍

File: swap_meet/vendor.py [hunk 15]
```diff
@@ -1,2 +1,158 @@
...
+        '''Adds one_item to the instance's inventory list'''
+        self.inventory.append(one_item)
+        return one_item
+
+    def remove(self, one_item):
```
Comment (anselrognlie) [inline swap_meet/vendor.py:52]: 👍

File: swap_meet/vendor.py [hunk 16]
```diff
@@ -1,2 +1,158 @@
...
+            self.inventory.remove(one_item)
+            return one_item
+        except ValueError:
+            return False
+        
+    def get_by_category(self, category_to_check):
+        '''
+        Takes a cateogry and returns a list of the items currently in the instance's inventory with that category attribute
+        '''
+        list_by_cat = [item for item in self.inventory if item.category == category_to_check]
```
Comment (anselrognlie) [inline swap_meet/vendor.py:66]: Great use of a list comprehension to get this result!

File: swap_meet/vendor.py [hunk 17]
```diff
@@ -1,2 +1,158 @@
...
+        their_item : Item instance
+            the item from swap_friend's inventory to swap
+
+        ---Returns---
+        Returns true if both items are in their respective inventories and the items were switched, otherwise returns False
+        '''
+        if my_item not in self.inventory:
+            return False
+        if their_item not in swap_friend.inventory:
+            return False
```
Comment (anselrognlie) [inline swap_meet/vendor.py:87]: I notice that you switched from a more EAFP approach in `remove` to a LBYL approach here. This is a good call. In `remove` there's only one thing that can fail, and it either happened or not. Here, there are two removes that could fail. And if the second failed, we would need to do work to undo the first operation that didn't fail. So here, checking that the everything will work successfully ahead of winds up being a more manageable approach.

File: swap_meet/vendor.py [hunk 18]
```diff
@@ -1,2 +1,158 @@
...
+
+    def swap_first_item(self, swap_friend):
+        '''
+        Takes the first item of the current instance's inventory and the first item of swap_friend's inventory and swaps them
+        '''
+        if not self.inventory or not swap_friend.inventory:
+            return False
+        my_item = self.inventory[0]
+        their_item = swap_friend.inventory[0]
+        self.swap_items(swap_friend, my_item, their_item)
```
Comment (anselrognlie) [inline swap_meet/vendor.py:102]: Nice reuse of your `swap_items`. Notice that due to the remove operations in `swap_items`, it is a linear operation, regardless of where in the list each item is. We could achieve constant time here either by swapping the items directly, or by doing an `index` then swap approach in `swap_items`.

File: swap_meet/vendor.py [hunk 19]
```diff
@@ -1,2 +1,158 @@
...
+        return True
+        
+    def get_best_by_category(self, category):
+        '''
+        Searches an instance's inventory attribute for an item of a specific category with the highest condition and returns it
+        '''
+        category_list = self.get_by_category(category)
+        if not category_list:
+            return None
+        return max(category_list, key=lambda x: x.condition)
```
Comment (anselrognlie) [inline swap_meet/vendor.py:112]: You've used `max()` a few times by now, along with the a custom key function. Do you think you could write your own compatible `max()` implementation? Give it a shot! How do you accept a function as an argument to another function and call it (think back to decorators)?

File: swap_meet/vendor.py [hunk 20]
```diff
@@ -1,2 +1,158 @@
...
+            the category of item to be searched for in the current instance's inventory  
+
+        ---Returns---
+        Returns False if there were no items of category my_priority in other's inventory or if there were no items of category their_priority in the current instance's inventory. Returns true if these exist and were swapped.
+        '''
+        to_trade = self.get_best_by_category(their_priority)
+        to_receive = other.get_best_by_category(my_priority)
+        if not to_trade or not to_receive:
+            return False
+        return self.swap_items(other, to_trade, to_receive)
```
Comment (anselrognlie) [inline swap_meet/vendor.py:133]: Nice clean implementation and function reuse!

File: swap_meet/vendor.py [hunk 21]
```diff
@@ -1,2 +1,158 @@
...
+    def get_newest_item(self) -> Item:
+        '''
+        Searches an instance's inventory attribute for an item with the youngest age. If inventory is empty, returns none.
+        '''
+        if not self.inventory:
+            return None
+        newest_item = self.inventory[0]
+        for item in self.inventory:
+            if item.age is None:
+                return None
```
Comment (anselrognlie) [inline swap_meet/vendor.py:144]: This is absolutely one possible approach to this function. We might also take the view that we could find the newest item that _has_ an age. But nicely done!

File: swap_meet/vendor.py [hunk 22]
```diff
@@ -1,2 +1,158 @@
...
+            elif item.age < newest_item.age:
+                newest_item = item
+        return newest_item
+    
+    def swap_by_newest(self, swap_friend):
```
Comment (anselrognlie) [inline swap_meet/vendor.py:149]: 👍 
Great reuse of your custom function along with the swap function.

File: swap_meet/item.py [hunk 23]
```diff
@@ -1,2 +1,69 @@
...
+        age : int or float, optional
+            None by default, otherwise a number representing how old the item is in years
+        '''
+        self.category = category
+        self.condition = condition
+        self.age = age
+
+
+    def __str__(self):
+        '''sets dunder str to a string'''
```
Comment (anselrognlie) [inline swap_meet/item.py:41]: I was curious what some of the built-in classes report for their `__str__`. It turns out "Return str(self)." 😆

File: swap_meet/item.py [hunk 24]
```diff
@@ -1,2 +1,69 @@
...
+        '''
+        descriptors = [
+            "Mint",
+            "Like New",
+            "Gently Used",
+            "Used",
+            "Well Used",
+            "Bad Condition",
+        ]
+        if 1 > self.condition >= 0:
```
Comment (anselrognlie) [inline swap_meet/item.py:56]: Love that you put the descriptions in a list (data-driven). Once you do that, try to write an approach that doesn't require essentially using each position as an individual hard-coded value. For instance, if we added an additional condition, we would have to add additional new clauses in the if chain to check for the new condition. What if instead we could either loop through the list, or somehow otherwise convert the condition into an index we could use?

File: swap_meet/item.py [hunk 25]
```diff
@@ -1,2 +1,69 @@
...
+        elif 3 > self.condition >= 2:
+            return descriptors[2]
+        elif 4 > self.condition >= 3:
+            return descriptors[3]
+        elif 5 > self.condition >= 4:
+            return descriptors[4]
+        elif self.condition == 5:
+            return descriptors[5]
+        else:
+            return f"Condition rated as {self.condition}"
```
Comment (anselrognlie) [inline swap_meet/item.py:69]: I really like this default clause!

File: swap_meet/clothing.py [hunk 26]
```diff
@@ -1,2 +1,34 @@
-class Clothing:
-    pass
\ No newline at end of file
+from swap_meet.item import Item
+class Clothing(Item):
```
Comment (anselrognlie) [inline swap_meet/clothing.py:2]: Definitely need to import `Item` so that we can use it as the parent class to `Clothing`.

File: swap_meet/clothing.py [hunk 27]
```diff
@@ -1,2 +1,34 @@
...
+    '''
+    def __init__(self, condition=0, age=None):
+        '''
+        Constructs all the necessary attributes for a clothing object; inherits from Item
+
+        ---Paramters---
+        condition : int or float, optional
+            zero by default, otherwise a number representing the condition of the item
+        '''
+        super().__init__("Clothing", condition, age)
```
Comment (anselrognlie) [inline swap_meet/clothing.py:29]: Nice job  removing the category from the constructor, since we don't want to encourager the caller to try to change it. Instead, we pass the literal value that we want to set up to the parent's constructor. We could also choose to use named parameters as
```python
        super().__init__(category="Clothing", condition=condition, age=age)
```

**Review by anselrognlie — COMMENTED** [review]
Nice job!

All required waves (and then some!) are complete and all tests are passing (including the tests which need completion)! Your code is well organized, and you are using good, descriptive variable names. You had great code reuse, and appear well on your way with OOP.

I don't see a lot to critique here. Keep thinking about code complexity. Keep making good use of helper functions (whether previous functions that can be re-used, or custom helpers). I also liked to see you heading in a data-driven direction on the condition strings. Push a little more there and I think you'll have a function that could work for any arbitrary collection of condition strings!

Well done!
